Importing necessary modules 

In [1]:
import tensorflow as tf
from tensorflow.keras import datasets, layers, models, losses
from sklearn.metrics import f1_score
import numpy as np

Loading the mnist classification data

In [2]:
(x_train,y_train),(x_test,y_test) = datasets.mnist.load_data()
x_train.shape

(60000, 28, 28)

Padding the dataset

In [3]:
x_train = tf.pad(x_train, [[0, 0], [2,2], [2,2]])/255
x_test = tf.pad(x_test, [[0, 0], [2,2], [2,2]])/255
x_train.shape

TensorShape([60000, 32, 32])

Expanding the tensor dimensions

In [4]:
x_train = tf.expand_dims(x_train, axis=3, name=None)
x_test = tf.expand_dims(x_test, axis=3, name=None)
x_train.shape

TensorShape([60000, 32, 32, 1])

Separating for validation set

In [5]:
x_val = x_train[-2000:,:,:,:] 
y_val = y_train[-2000:] 
x_train = x_train[:-2000,:,:,:] 
y_train = y_train[:-2000]

Building the model by changing Activation function from Dying Relu to tanh 

In [6]:
model = models.Sequential()

model.add(layers.Conv2D(6, 5, activation='tanh', input_shape=x_train.shape[1:]))
model.add(layers.AveragePooling2D(2))
model.add(layers.Activation('sigmoid'))

model.add(layers.Conv2D(16, 5, activation='tanh'))
model.add(layers.AveragePooling2D(2))
model.add(layers.Activation('sigmoid'))

model.add(layers.Conv2D(120, 5, activation='tanh'))

model.add(layers.Flatten())

model.add(layers.Dense(84, activation='tanh'))
model.add(layers.Dense(10, activation='softmax'))

model.summary()

c:\Users\sneha\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 28, 28, 6)      │           156 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ average_pooling2d               │ (None, 14, 14, 6)      │             0 │
│ (AveragePooling2D)              │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ activation (Activation)         │ (None, 14, 14, 6)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 10, 10, 16)     │         2,416 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ average_pooling2d_1             │ (None, 5, 5, 16)       │             0 │
│ (AveragePooling2D)              │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ activation_1 (Activation)       │ (None, 5, 5, 16)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_2 (Conv2D)               │ (None, 1, 1, 120)      │        48,120 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 120)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 84)             │        10,164 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │           850 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 61,706 (241.04 KB)

 Trainable params: 61,706 (241.04 KB)

 Non-trainable params: 0 (0.00 B)

Compiling model with Adam optimizer for 40 epochs

In [7]:
model.compile(optimizer='adam', loss=losses.sparse_categorical_crossentropy, metrics=['accuracy'])
history = model.fit(x_train, y_train, batch_size=64, epochs=40, validation_data=(x_val, y_val))

Epoch 1/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.2576 - loss: 1.9868 - val_accuracy: 0.8915 - val_loss: 0.3936
Epoch 2/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.8572 - loss: 0.4475 - val_accuracy: 0.9315 - val_loss: 0.2369
Epoch 3/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.9012 - loss: 0.3119 - val_accuracy: 0.9305 - val_loss: 0.2468
Epoch 4/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.9164 - loss: 0.2605 - val_accuracy: 0.9570 - val_loss: 0.1518
Epoch 5/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.9267 - loss: 0.2302 - val_accuracy: 0.9680 - val_loss: 0.1211
Epoch 6/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9366 - loss: 0.1951 - val_accuracy: 0.9530 - val_loss: 0.1553
Epoch 7/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9410 - loss: 0.1816 - val_accuracy: 0.9580 - val_loss: 0.1400
Epoch 8/40
907/907 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.9455 - loss: 0.1694 - val_accuracy: 0.

Model's accuracy

In [8]:
model.evaluate(x_test, y_test)
print(f"Accuracy on test set: {model.evaluate(x_test, y_test)[1]:.2f}")

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9737 - loss: 0.0855
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9737 - loss: 0.0855
Accuracy on test set: 0.98


F1 SCORE

F1 Score Range: The F1 score ranges from 0 to 1:

In [9]:
predictions = model.predict(x_test)

predicted_classes = np.argmax(predictions, axis=1)

f1 = f1_score(y_test, predicted_classes, average='macro')

print(f"f1 score: {f1:.2f}")

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step
f1 score: 0.98


After changing number of epochs:

    The accuracy here is 0.98

    The F1 Score here is 0.98